crates/ra_mbe/src/syntax_bridge.rs

```diff
@@ -321,40 +322,51 @@ impl RawConvertor<'_> {
 
             if let Some((kind, closed)) = delim {
                 let mut subtree = tt::Subtree::default();
-                let id = self.id_alloc.open_delim(range);
+                let id = self.id_alloc().open_delim(range);
                 subtree.delimiter = Some(tt::Delimiter { kind, id });
```

Comment:
This delimiter seems to only be used in the case where we find the closing delim. I think this line, as well as the one above which allocates an ID for the open delim could be moved to below the `let last_range = match ..` block. This would mean we wouldn't need to remove this ID from the token map in the `None` arm of that match statement with `self.id_alloc().close_delim(id, None)`. 

I haven't looked super closely, but I think this may even allow removing the `open_delim` and `close_delim` methods entirely from `TokenIdAlloc` in favor of just the `delim` method, since we would only be creating them in pairs and below the match statement we'd already have the `last_range` for the closing delim. 

I think this approach would be cleaner than optimistically creating the delimiter and then cleaning it up if we don't find the closing delim. 

I'm not too familiar with this codebase, so let me know if something I said here is off base.

Reply:
Nice catch !

The reason to alloc the delimiter ID like that is twofold:

1.  We want to order the `TokenId` in a way that ancestors have a smaller ID than descendants, which make searching the TokenID in HIR faster.   
2. And the *REAL* reason is: there are tons of tests depends on the TokenID order, if we I changed the order of TokenId, I have to fix these. I would like to separate it to another PR.